<a href="https://colab.research.google.com/github/Saramendez12/Entrega_Proyectos/blob/main/Proyecto3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

En el presente proyecto se realizara una limpieza de datos a la tabla entregada  ```
dataset.1
```
la cual es una tabla con las siguientes variables, una identificación de usuario, genero, salario, si es comprador o no y fecha de nacimiento.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


**Importación de Librerías** 



In [ ]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt

**Carga de Datos**

In [ ]:
from google.colab import drive
drive.mount('/content/drive/')

Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount("/content/drive/", force_remount=True).


In [ ]:
data=pd.read_csv("/content/dataset_1.csv")

Se realiza una visualización inicial de los datos que nos brinda información de la identificación de usuario, genero, salario, si es comprador o no y fecha de nacimiento,tambien el tamaño del Dataframe que representa el numero de usuarios y sus respectivos datos existentes en el DataFrame.

In [ ]:
data

user_id  gender      wage  purchased   birth_day
0         217  Female       NaN          0  1983-08-06
1         226  Female  146500.0          1  1972-01-28
2         697    Male       NaN          1  1981-12-12
3         743  Female       NaN          0  1983-05-26
4          63  Female  140500.0          1  1967-11-02
...       ...     ...       ...        ...         ...
1169      636    Male   76500.0          0  1970-08-26
1170      320  Female  143500.0          1  1970-04-07
1171      136  Female       NaN          1  1968-12-28
1172      721    Male  148000.0          1  1993-11-23
1173      471     man   60500.0          0  1984-06-02

[1174 rows x 5 columns]

**Exploración Inicial**

Se realiza una exploración de las 5 primeras filas del Dataframe. En la cual observamos a simple vista la existencia de datos faltantes(NaN) dentro de la columna wage.

In [ ]:
data.head()

user_id  gender      wage  purchased   birth_day
0      217  Female       NaN          0  1983-08-06
1      226  Female  146500.0          1  1972-01-28
2      697    Male       NaN          1  1981-12-12
3      743  Female       NaN          0  1983-05-26
4       63  Female  140500.0          1  1967-11-02

Se realiza una exploración de las 5 ultimas filas del Dataframe. En la cual nos permite entender que existen 1173 indices dentro del Dataframe.

In [ ]:
data.tail()

user_id  gender      wage  purchased   birth_day
1169      636    Male   76500.0          0  1970-08-26
1170      320  Female  143500.0          1  1970-04-07
1171      136  Female       NaN          1  1968-12-28
1172      721    Male  148000.0          1  1993-11-23
1173      471     man   60500.0          0  1984-06-02

Se presenta la información general del Data Frame , en cual podemos encontrar que se maneja un rango de indice de 0 a 1173, 5 columnas para cada entrada que en total son 1174. A partir de esto observamos que la cantidad de datos no nulos por cada variable dandonos a entender una vez más que la variable wage tiene datos faltantes.Podemos analizar el tipo de dato de cada columna asociado a una variable, en este caso presentamos tipos de datos float(Decimales),int (Enteros) y object(Objeto).Y por ultimo la información nos arroja la cantidad de memoria usada por el DataFrame.

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1174 entries, 0 to 1173
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   user_id    1174 non-null   int64  
 1   gender     1174 non-null   object 
 2   wage       930 non-null    float64
 3   purchased  1174 non-null   int64  
 4   birth_day  1174 non-null   object 
dtypes: float64(1), int64(2), object(2)
memory usage: 46.0+ KB


A partir del metodo describe nos permite generar las estadísticas descriptivas para nuestro DataFrame, estas incluyen un conteo del numero de datos por variable (count), una media (mean), una desviación estandar(std), el valor minimo de los datos (min), los cuartiles(1,2,3) y el valor maximo de los datos(max). 

Tomando en cuenta esta información podemos afirmar que se presenta dentro de la variable wage un sesgo negativo dado que el valor de la mediana (71250.0) es menor al valor de la media (71948.39). Por ultimo recordar que la información de la variable purchased es un bool (1 o 0/si o no) por lo que las estadisticas describas no aportan mucha información útil.

In [ ]:
data.describe()

user_id           wage    purchased
count  1174.000000     930.000000  1174.000000
mean    499.988927   71948.387097     0.404600
std     287.648865   34112.616989     0.491024
min       1.000000   15000.000000     0.000000
25%     252.250000   45500.000000     0.000000
50%     499.500000   71250.000000     0.000000
75%     749.750000   89500.000000     1.000000
max    1000.000000  152500.000000     1.000000

**Revisión de duplicados Explícitos**

Validamos lo visto en la exploración inicial, para esto primero comprobaremos si se encuentran registros duplicados y si es así nos dira cuantos hay.

In [ ]:
data.duplicated().sum()

174

Eliminación de duplicados explícitos
-Se realiza la eliminación de los datos duplicados en el DataFrame, a partir de esto se reinicia los indices y se actualiza nuestro data.

In [ ]:
data=data.drop_duplicates().reset_index(drop=True)

En la actualización del datos sin los datos duplicados observamos que ahora tenemos 1000 entradas.

In [ ]:
data

user_id  gender      wage  purchased   birth_day
0        217  Female       NaN          0  1983-08-06
1        226  Female  146500.0          1  1972-01-28
2        697    Male       NaN          1  1981-12-12
3        743  Female       NaN          0  1983-05-26
4         63  Female  140500.0          1  1967-11-02
..       ...     ...       ...        ...         ...
995      305  Female   22500.0          1  1969-03-09
996      636    Male   76500.0          0  1970-08-26
997      320  Female  143500.0          1  1970-04-07
998      721    Male  148000.0          1  1993-11-23
999      471     man   60500.0          0  1984-06-02

[1000 rows x 5 columns]

**Revisión de duplicados implícitos**
Encontramos que en la columna de gender se encuentran 2 datos duplicados implicitos los cuales son male,man,woman y female. Todos los demás datos son unicos.

In [ ]:
for i in data.columns:
  print(i)
  print(data[i].nunique())
  print(data[i].unique())

user_id
1000
[ 217  226  697  743   63   92  971  777  753  211  891  817  587  889
  126  238  733  483  463  139  994  473  316  662  578  180  631  129
  688  341  382   54  998  961  583  128  542  654  950  367  243  488
   80  307  793  269  607  319  702  903  745    1  679  757  689  440
  568  922  701   64  346  775  432  551  107  674  283  321  653  393
  962  357   73  771  304  429  703   69  363  308  544  419  599  917
  744  555  867   82  120  175  279  329  567  188  833  579  859  142
  791  589  276  704   19  729  815  651  850  335  200  944  947  621
  204  780  902  417  420  462  750  273  522  907  360  968  672  262
  324  881  285  371  554  720  250  151  914  625  862  106  582  192
  766  687  983  841  242  479  228  478  178  206  616  931  286  788
  378   23  369  428  513  541  433  476   90  127  834  553  386  915
  822   25  540  957  987  213  591   55  742  698  425  576  893  796
  974  160  100  581  716  300  695   27  911  942  933  533  69

Se realiza el remplazo de los datos implicitos dentro de la columna gender. Dejando solo Female y Male como datos unicos dentro del data.

In [ ]:
data.replace({"woman":"Female","man":"Male"}, inplace=True)

In [ ]:
data

user_id  gender      wage  purchased   birth_day
0        217  Female       NaN          0  1983-08-06
1        226  Female  146500.0          1  1972-01-28
2        697    Male       NaN          1  1981-12-12
3        743  Female       NaN          0  1983-05-26
4         63  Female  140500.0          1  1967-11-02
..       ...     ...       ...        ...         ...
995      305  Female   22500.0          1  1969-03-09
996      636    Male   76500.0          0  1970-08-26
997      320  Female  143500.0          1  1970-04-07
998      721    Male  148000.0          1  1993-11-23
999      471    Male   60500.0          0  1984-06-02

[1000 rows x 5 columns]

Se realiza una verificación de que ya no hayan datos duplicados explicitos.

In [ ]:
data.duplicated().sum()

0

Se realiza una verificación de que ya no hayan datos duplicados implicitos, los cuales anteriormente se encontraban en la columna gender.

In [ ]:
for i in data.columns:
  print(i)
  print(data[i].nunique())
  print(data[i].unique())

user_id
1000
[ 217  226  697  743   63   92  971  777  753  211  891  817  587  889
  126  238  733  483  463  139  994  473  316  662  578  180  631  129
  688  341  382   54  998  961  583  128  542  654  950  367  243  488
   80  307  793  269  607  319  702  903  745    1  679  757  689  440
  568  922  701   64  346  775  432  551  107  674  283  321  653  393
  962  357   73  771  304  429  703   69  363  308  544  419  599  917
  744  555  867   82  120  175  279  329  567  188  833  579  859  142
  791  589  276  704   19  729  815  651  850  335  200  944  947  621
  204  780  902  417  420  462  750  273  522  907  360  968  672  262
  324  881  285  371  554  720  250  151  914  625  862  106  582  192
  766  687  983  841  242  479  228  478  178  206  616  931  286  788
  378   23  369  428  513  541  433  476   90  127  834  553  386  915
  822   25  540  957  987  213  591   55  742  698  425  576  893  796
  974  160  100  581  716  300  695   27  911  942  933  533  69

**Transformación de tipos de datos**
Nuestra columna de la variable gender tenia un tipo de dato de objeto, realizamos una transformación a que sea un tipo de dato categorico.

In [ ]:
data = data.astype({"gender":"category","purchased":"bool"})

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype   
---  ------     --------------  -----   
 0   user_id    1000 non-null   int64   
 1   gender     1000 non-null   category
 2   wage       786 non-null    float64 
 3   purchased  1000 non-null   bool    
 4   birth_day  1000 non-null   object  
dtypes: bool(1), category(1), float64(1), int64(1), object(1)
memory usage: 25.6+ KB


**Enriquecimiento del dataset**
*   Se crea una variable con los años o edad de los individuos y se crea ua nueva columna con estos datos.



In [ ]:
data["birth_day"]=pd.to_datetime(data["birth_day"],format="%Y-%m-%d")

In [ ]:
data["age"]=((pd.to_datetime('2022/09/2', format="%Y-%m-%d"))-data["birth_day"]).dt.days

In [ ]:
data['age'] = data['age']/365.2425
data["age"]= data['age'].astype(int)

Hacemos una verificación de la creación de nuestro DataFrame

In [ ]:
data

user_id  gender      wage  purchased  birth_day  age
0        217  Female       NaN      False 1983-08-06   39
1        226  Female  146500.0       True 1972-01-28   50
2        697    Male       NaN       True 1981-12-12   40
3        743  Female       NaN      False 1983-05-26   39
4         63  Female  140500.0       True 1967-11-02   54
..       ...     ...       ...        ...        ...  ...
995      305  Female   22500.0       True 1969-03-09   53
996      636    Male   76500.0      False 1970-08-26   52
997      320  Female  143500.0       True 1970-04-07   52
998      721    Male  148000.0       True 1993-11-23   28
999      471    Male   60500.0      False 1984-06-02   38

[1000 rows x 6 columns]

Realizando una nueva exploración de las estadisiticas descriptivas de las variables encontramos dentro de la variable wage un sesgo positivo dado que el valor de la mediana (71750.0) es menor al valor de la media (72580.15). Así mismo hacemos nuestro primer analisis de la variable age ya creada, observamos un sesgo positivo de igual manera ya que la mediana es menor a la edad y una media de 40 años aproximados

In [ ]:
data.describe()

user_id           wage         age
count  1000.000000     786.000000  1000.00000
mean    500.500000   72580.152672    39.73700
std     288.819436   34268.329521    10.72151
min       1.000000   15000.000000    17.00000
25%     250.750000   46125.000000    32.00000
50%     500.500000   71750.000000    39.00000
75%     750.250000   90000.000000    48.00000
max    1000.000000  152500.000000    63.00000

**Verificación de missing values**

Se realiza una verificación de el numero de datos faltantes dentro del Dataframe. 
Tal como se habia mencionado y analizado en las exploraciones iniciales la variable wage tiene datos faltantes en este caso presenta 214 faltantes.

In [ ]:
data.isna().sum()

user_id        0
gender         0
wage         214
purchased      0
birth_day      0
age            0
dtype: int64

 **Tratamiento de los missing values
(Eliminación o rellenado de Valores Faltantes)**

En el tratamiento de los datos faltantes, realizamos dos imputaciones para la primera ponemos el valor de una media general en los faltantes y para la segunda ponemos el valor de ua media condicionada por grupos en los faltaste haciendo esto el tratamiento o manipulación de datos faltantes de forma más robusta.

Primera imputación : Rellenado de los datos nulos por el valor de la media lo cual no es muy robusto



In [ ]:
data_i= data["wage"].fillna(data["wage"].mean())
data_i

0       72580.152672
1      146500.000000
2       72580.152672
3       72580.152672
4      140500.000000
           ...      
995     22500.000000
996     76500.000000
997    143500.000000
998    148000.000000
999     60500.000000
Name: wage, Length: 1000, dtype: float64

Se realiza una nueva exploración inicial de los datos despues de la primera imputación, en la cual se observa que hay una distribución simetrica y que en el conteo de datos ya se encuentran las 1000 entradas completas dentro de la variable wage. Así mismo como se relleno con la media esta permanece en su valor. Pero lo más importante el problema de datos nulos es arreglado. 

In [ ]:
data_i.describe()

count      1000.000000
mean      72580.152672
std       30377.009614
min       15000.000000
25%       53500.000000
50%       72580.152672
75%       83000.000000
max      152500.000000
Name: wage, dtype: float64

Realizamos la verificación de eliminación de los datos faltantes dentro del data.

In [ ]:
data_i.isna().sum()

0

Rellenado de los datos nulos por el valor de la mediana lo cual no es muy robusto

In [ ]:
data_i2= data["wage"].fillna(data["wage"].median())
data_i2

0       71750.0
1      146500.0
2       71750.0
3       71750.0
4      140500.0
         ...   
995     22500.0
996     76500.0
997    143500.0
998    148000.0
999     60500.0
Name: wage, Length: 1000, dtype: float64

In [ ]:
data_i2.describe()

count      1000.000000
mean      72402.500000
std       30378.919459
min       15000.000000
25%       53500.000000
50%       71750.000000
75%       83000.000000
max      152500.000000
Name: wage, dtype: float64

In [ ]:
data_i2.isna().sum()

0

Segunda imputación : Inyectamos una media condicionada a los datos nulos a traves de las tablas pivot,debido  a que este metodo nos otorga un mayor grado de robustes que la imputación 1. Para realizar este proceso es necesario crear grupos etarios para poder verlo más detallado.

In [ ]:
data

user_id  gender      wage  purchased  birth_day  age
0        217  Female       NaN      False 1983-08-06   39
1        226  Female  146500.0       True 1972-01-28   50
2        697    Male       NaN       True 1981-12-12   40
3        743  Female       NaN      False 1983-05-26   39
4         63  Female  140500.0       True 1967-11-02   54
..       ...     ...       ...        ...        ...  ...
995      305  Female   22500.0       True 1969-03-09   53
996      636    Male   76500.0      False 1970-08-26   52
997      320  Female  143500.0       True 1970-04-07   52
998      721    Male  148000.0       True 1993-11-23   28
999      471    Male   60500.0      False 1984-06-02   38

[1000 rows x 6 columns]

In [ ]:
data["age_2"] = pd.cut(data["age"],4)

Cremos una tabla pivot  que muestre la media del salario a partir del grupo etario y el sexo 


In [ ]:
pivot=data.pivot_table(index=["age_2","gender"],values="wage",aggfunc="mean")
pivot

wage
age_2          gender              
(16.954, 28.5] Female  57826.388889
               Male    63843.283582
(28.5, 40.0]   Female  76953.846154
               Male    71471.518987
(40.0, 51.5]   Female  78301.526718
               Male    68646.788991
(51.5, 63.0]   Female  77168.831169
               Male    86392.857143

Creamos una función que nos permite verificar los datos faltantes y hacer el reemplazo.

In [ ]:
def funcion(row):
  wage=row["wage"]
  age=row["age_2"]
  gender=row["gender"]
  if pd.isna(wage):
    return pivot["wage"][age][gender]
  return wage

Realizamos el llenado de la media condicional sobre nuestro base de dato original

In [ ]:
data["wage"]=data.apply(funcion,axis=1)

In [ ]:
data

user_id  gender           wage  purchased  birth_day  age           age_2
0        217  Female   76953.846154      False 1983-08-06   39    (28.5, 40.0]
1        226  Female  146500.000000       True 1972-01-28   50    (40.0, 51.5]
2        697    Male   71471.518987       True 1981-12-12   40    (28.5, 40.0]
3        743  Female   76953.846154      False 1983-05-26   39    (28.5, 40.0]
4         63  Female  140500.000000       True 1967-11-02   54    (51.5, 63.0]
..       ...     ...            ...        ...        ...  ...             ...
995      305  Female   22500.000000       True 1969-03-09   53    (51.5, 63.0]
996      636    Male   76500.000000      False 1970-08-26   52    (51.5, 63.0]
997      320  Female  143500.000000       True 1970-04-07   52    (51.5, 63.0]
998      721    Male  148000.000000       True 1993-11-23   28  (16.954, 28.5]
999      471    Male   60500.000000      False 1984-06-02   38    (28.5, 40.0]

[1000 rows x 7 columns]

Al crear la función verificamos que ya no existan datos faltantes.

In [ ]:
data.isna().sum()

user_id      0
gender       0
wage         0
purchased    0
birth_day    0
age          0
age_2        0
dtype: int64

**Outliders**
 Primero identificamos el primer cuartil, el tercer cuartil,el rango intercuartilico, y el limite superior e inferior de la distribucion de los datos de la variable wage 

In [ ]:
variable_wage=data.loc[::1,"wage"]

In [ ]:
variable_wage

0       76953.846154
1      146500.000000
2       71471.518987
3       76953.846154
4      140500.000000
           ...      
995     22500.000000
996     76500.000000
997    143500.000000
998    148000.000000
999     60500.000000
Name: wage, Length: 1000, dtype: float64

In [ ]:
variable_wage.describe()

count      1000.000000
mean      72576.657129
std       30553.189624
min       15000.000000
25%       53500.000000
50%       71500.000000
75%       84625.000000
max      152500.000000
Name: wage, dtype: float64

In [ ]:
Q_1=variable_wage.quantile(0.25)

In [ ]:
Q_1

53500.0

In [ ]:
Q_3=variable_wage.quantile(0.75)

In [ ]:
Q_3

84625.0

In [ ]:
rango_intercu=Q_3-Q_1

In [ ]:
rango_intercu


31125.0

In [ ]:
limite_sup=(rango_intercu*1.5)+Q_3
limite_sup

131312.5

In [ ]:
limite_inf=Q_1-(1.5*rango_intercu)
limite_inf

6812.5

Se determinan los datos que mayores al limite superior y se toma su posiion dentro del dataset 

In [ ]:
outlier=variable_wage[(variable_wage>limite_sup)]

In [ ]:
indice_out=outlier.index

In [ ]:
indice_out

Int64Index([  1,   4,   7,  15,  46,  70,  80,  89,  98, 130, 149, 161, 212,
            221, 251, 252, 261, 273, 274, 320, 345, 347, 360, 369, 375, 409,
            415, 495, 513, 517, 518, 522, 546, 576, 586, 587, 589, 599, 643,
            661, 666, 687, 691, 693, 714, 738, 740, 741, 778, 784, 792, 806,
            823, 824, 829, 836, 856, 864, 883, 886, 893, 895, 905, 908, 914,
            918, 945, 948, 950, 997, 998],
           dtype='int64')

Se eliminan los outliers

In [ ]:
no_outliers=variable_wage[~variable_wage.index.isin(indice_out)]

In [ ]:
no_outliers.describe()

count       929.000000
mean      67297.262787
std       24678.398307
min       15000.000000
25%       51500.000000
50%       71471.518987
75%       80000.000000
max      131000.000000
Name: wage, dtype: float64

In [ ]:
outliers_2=no_outliers[no_outliers<min([limite_inf])]
outliers_2

Series([], Name: wage, dtype: float64)

Se introduce la variable sin outliers dentro del dataframe  

In [ ]:
data_2=data.copy()

In [ ]:
data_2

user_id  gender           wage  purchased  birth_day  age           age_2
0        217  Female   76953.846154      False 1983-08-06   39    (28.5, 40.0]
1        226  Female  146500.000000       True 1972-01-28   50    (40.0, 51.5]
2        697    Male   71471.518987       True 1981-12-12   40    (28.5, 40.0]
3        743  Female   76953.846154      False 1983-05-26   39    (28.5, 40.0]
4         63  Female  140500.000000       True 1967-11-02   54    (51.5, 63.0]
..       ...     ...            ...        ...        ...  ...             ...
995      305  Female   22500.000000       True 1969-03-09   53    (51.5, 63.0]
996      636    Male   76500.000000      False 1970-08-26   52    (51.5, 63.0]
997      320  Female  143500.000000       True 1970-04-07   52    (51.5, 63.0]
998      721    Male  148000.000000       True 1993-11-23   28  (16.954, 28.5]
999      471    Male   60500.000000      False 1984-06-02   38    (28.5, 40.0]

[1000 rows x 7 columns]

In [ ]:
data_2["wage"]=no_outliers

In [ ]:
data_2

user_id  gender          wage  purchased  birth_day  age           age_2
0        217  Female  76953.846154      False 1983-08-06   39    (28.5, 40.0]
1        226  Female           NaN       True 1972-01-28   50    (40.0, 51.5]
2        697    Male  71471.518987       True 1981-12-12   40    (28.5, 40.0]
3        743  Female  76953.846154      False 1983-05-26   39    (28.5, 40.0]
4         63  Female           NaN       True 1967-11-02   54    (51.5, 63.0]
..       ...     ...           ...        ...        ...  ...             ...
995      305  Female  22500.000000       True 1969-03-09   53    (51.5, 63.0]
996      636    Male  76500.000000      False 1970-08-26   52    (51.5, 63.0]
997      320  Female           NaN       True 1970-04-07   52    (51.5, 63.0]
998      721    Male           NaN       True 1993-11-23   28  (16.954, 28.5]
999      471    Male  60500.000000      False 1984-06-02   38    (28.5, 40.0]

[1000 rows x 7 columns]

In [ ]:
data_2=data_2.dropna().reset_index(drop=True)

In [ ]:
data_2

user_id  gender          wage  purchased  birth_day  age           age_2
0        217  Female  76953.846154      False 1983-08-06   39    (28.5, 40.0]
1        697    Male  71471.518987       True 1981-12-12   40    (28.5, 40.0]
2        743  Female  76953.846154      False 1983-05-26   39    (28.5, 40.0]
3         92  Female  33500.000000      False 1997-02-26   25  (16.954, 28.5]
4        971    Male  56500.000000      False 1997-07-23   25  (16.954, 28.5]
..       ...     ...           ...        ...        ...  ...             ...
924      600  Female  62500.000000      False 1982-02-16   40    (28.5, 40.0]
925      366    Male  68646.788991      False 1977-10-28   44    (40.0, 51.5]
926      305  Female  22500.000000       True 1969-03-09   53    (51.5, 63.0]
927      636    Male  76500.000000      False 1970-08-26   52    (51.5, 63.0]
928      471    Male  60500.000000      False 1984-06-02   38    (28.5, 40.0]

[929 rows x 7 columns]

In [ ]:
data_2.describe()

user_id           wage         age
count   929.000000     929.000000  929.000000
mean    501.013994   67297.262787   39.292788
std     287.993648   24678.398307   10.726034
min       1.000000   15000.000000   17.000000
25%     254.000000   51500.000000   31.000000
50%     500.000000   71471.518987   39.000000
75%     749.000000   80000.000000   47.000000
max    1000.000000  131000.000000   63.000000

In [ ]:
data.describe()

user_id           wage         age
count  1000.000000    1000.000000  1000.00000
mean    500.500000   72576.657129    39.73700
std     288.819436   30553.189624    10.72151
min       1.000000   15000.000000    17.00000
25%     250.750000   53500.000000    32.00000
50%     500.500000   71500.000000    39.00000
75%     750.250000   84625.000000    48.00000
max    1000.000000  152500.000000    63.00000

Eliminando los outliers de la variable salario encontramos que ahora la diferencia enre la media y la median es mayor, por lo que el valor de la asimeria negativa es mayor, tambien el valor de la media es menor cuando no hay outliers 

In [ ]:
data["unos"]=1

In [ ]:
data

user_id  gender           wage  purchased  birth_day  age  \
0        217  Female   76953.846154      False 1983-08-06   39   
1        226  Female  146500.000000       True 1972-01-28   50   
2        697    Male   71471.518987       True 1981-12-12   40   
3        743  Female   76953.846154      False 1983-05-26   39   
4         63  Female  140500.000000       True 1967-11-02   54   
..       ...     ...            ...        ...        ...  ...   
995      305  Female   22500.000000       True 1969-03-09   53   
996      636    Male   76500.000000      False 1970-08-26   52   
997      320  Female  143500.000000       True 1970-04-07   52   
998      721    Male  148000.000000       True 1993-11-23   28   
999      471    Male   60500.000000      False 1984-06-02   38   

              age_2  unos  
0      (28.5, 40.0]     1  
1      (40.0, 51.5]     1  
2      (28.5, 40.0]     1  
3      (28.5, 40.0]     1  
4      (51.5, 63.0]     1  
..              ...   ...  
995    (51.5, 63.0]     1  
996    (51.5, 63.0]     1  
997    (51.5, 63.0]     1  
998  (16.954, 28.5]     1  
999    (28.5, 40.0]     1  

[1000 rows x 8 columns]

In [ ]:
pd.pivot_table(data,index="age_2",columns="purchased",values="unos",aggfunc="sum")

purchased       False  True
age_2                      
(16.954, 28.5]    172     7
(28.5, 40.0]      286    87
(40.0, 51.5]      128   169
(51.5, 63.0]       12   139

**Manejo de memoria**

In [ ]:
data.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   user_id    1000 non-null   int64         
 1   gender     1000 non-null   category      
 2   wage       1000 non-null   float64       
 3   purchased  1000 non-null   bool          
 4   birth_day  1000 non-null   datetime64[ns]
 5   age        1000 non-null   int64         
 6   age_2      1000 non-null   category      
 7   unos       1000 non-null   int64         
dtypes: bool(1), category(2), datetime64[ns](1), float64(1), int64(3)
memory usage: 42.6 KB


In [ ]:
data.astype({"user_id":np.int32},{"wage":np.float32}).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   user_id    1000 non-null   int32         
 1   gender     1000 non-null   category      
 2   wage       1000 non-null   float64       
 3   purchased  1000 non-null   bool          
 4   birth_day  1000 non-null   datetime64[ns]
 5   age        1000 non-null   int64         
 6   age_2      1000 non-null   category      
 7   unos       1000 non-null   int64         
dtypes: bool(1), category(2), datetime64[ns](1), float64(1), int32(1), int64(2)
memory usage: 38.6 KB


**Datos inconsistentes**


Podemos deducir que no existen datos inconsistentes, ya que al momento de generar agrupaciones con determinadas variables se pudo apreciar que los datos mantenian un solo tipo de dato, ademas de que en las estadisticas descriptivas, realizadas a lo largo del proyecto, no se refleja que existieran datos de este tipo, ya que no exponian nigun tipo de error.

**Clasificación de Datos:**
   Categorización en grupos etarios

In [ ]:
data["age_2"] = pd.cut(data["age"],4) 

**Clasificación de Datos:**
   Categorización en grupos de ingreso

In [ ]:
data["intervalos wage"] = pd.cut(data["wage"],4)

**Tablas Pivot:**

Porcentaje de compras por grupos etarios

In [ ]:
Compras_etarios=pd.pivot_table(data,index="age_2",columns="purchased",values="unos",aggfunc="sum")
Compras_etarios

purchased       False  True
age_2                      
(16.954, 28.5]    172     7
(28.5, 40.0]      286    87
(40.0, 51.5]      128   169
(51.5, 63.0]       12   139

In [ ]:
Compras_etarios["suma"]=Compras_etarios[False]+Compras_etarios[True]

In [ ]:
Compras_etarios["% de compra"]=(Compras_etarios[True]/Compras_etarios["suma"])*100

In [ ]:
Compras_etarios["% de  no compra"]=(Compras_etarios[False]/Compras_etarios["suma"])*100

In [ ]:
Compras_etarios

purchased       False  True  suma  % de compra  % de  no compra
age_2                                                          
(16.954, 28.5]    172     7   179     3.910615        96.089385
(28.5, 40.0]      286    87   373    23.324397        76.675603
(40.0, 51.5]      128   169   297    56.902357        43.097643
(51.5, 63.0]       12   139   151    92.052980         7.947020

Con respecto al porcentaje de compras por grupo etario podemos obsarvar que en los 2 grupos etarios que poseen la poblacion más joven ([ 16.954, 28.5] y [28.5,40]) es más probable encontrar individuos que no hayan comprado. Algo que contrasta con los 2 grupos de mayor edad , en los que es más probable encontrar individuos que hayan comprado el producto


Porcentaje de compras por grupos de ingreso

In [ ]:
Compras_ingreso=data.pivot_table(index="intervalos wage", columns="purchased",values="unos",aggfunc="sum")
Compras_ingreso

purchased             False  True
intervalos wage                  
(14862.5, 49375.0]      118    96
(49375.0, 83750.0]      417   112
(83750.0, 118125.0]      53   106
(118125.0, 152500.0]     10    88

In [ ]:
Compras_ingreso["suma"]=Compras_ingreso[False]+Compras_ingreso[True]

In [ ]:
Compras_ingreso["% de compra"]=(Compras_ingreso[False]/Compras_ingreso["suma"])*100

In [ ]:
Compras_ingreso["% de no compra"]=(Compras_ingreso[True]/Compras_ingreso["suma"])*100

In [ ]:
Compras_ingreso

purchased             False  True  suma  % de compra  % de no compra
intervalos wage                                                     
(14862.5, 49375.0]      118    96   214    55.140187       44.859813
(49375.0, 83750.0]      417   112   529    78.827977       21.172023
(83750.0, 118125.0]      53   106   159    33.333333       66.666667
(118125.0, 152500.0]     10    88    98    10.204082       89.795918

Se ver que se nota una relacion entre menores ingresos con la compra, viendose esto reflejado en que los dos intervalos que representan una menor cantidad de ingreso son los intervalos donde el porcentaje de comprar es mayor que el procentaje de no comprar

Porcentaje compras por genero


In [ ]:
Compras_genero=data.pivot_table(index="gender", columns="purchased",values="unos",aggfunc="sum")
Compras_genero

purchased  False  True
gender                
Female       301   221
Male         297   181

In [ ]:
Compras_genero["suma"]=Compras_genero[False]+Compras_genero[True]

In [ ]:
Compras_genero["% de compra"]=(Compras_genero[False]/Compras_genero["suma"])*100

In [ ]:
Compras_genero["% de no compra"]=(Compras_genero[True]/Compras_genero["suma"])*100

In [ ]:
Compras_genero

purchased  False  True  suma  % de compra  % de no compra
gender                                                   
Female       301   221   522    57.662835       42.337165
Male         297   181   478    62.133891       37.866109

Acorde con lo mosrado en la tabla podemos decir que tanto hombres como mujeres son propensos a comprar, aunque el porcentaje de compra de los hombers es un poco mayor que el de las mujeres

Porcentaje de compras por grupos etarios, grupos de ingreso y género


In [ ]:
Compras_etinge=data.pivot_table(index=["age_2","intervalos wage","gender"], columns="purchased",values="unos",aggfunc="sum")
Compras_etinge

purchased                                   False  True
age_2          intervalos wage      gender             
(16.954, 28.5] (14862.5, 49375.0]   Female     28     0
                                    Male       20     0
               (49375.0, 83750.0]   Female     46     2
                                    Male       50     0
               (83750.0, 118125.0]  Female     12     1
                                    Male       15     0
               (118125.0, 152500.0] Female      1     1
                                    Male        0     3
(28.5, 40.0]   (14862.5, 49375.0]   Female     21     0
                                    Male       31     0
               (49375.0, 83750.0]   Female    104    13
                                    Male      112    17
               (83750.0, 118125.0]  Female      9     8
                                    Male        8    17
               (118125.0, 152500.0] Female      1    18
                                    Male        0    14
(40.0, 51.5]   (14862.5, 49375.0]   Female     11    27
                                    Male        4    29
               (49375.0, 83750.0]   Female     48    24
                                    Male       53    21
               (83750.0, 118125.0]  Female      6    19
                                    Male        0    20
               (118125.0, 152500.0] Female      4    21
                                    Male        2     8
(51.5, 63.0]   (14862.5, 49375.0]   Female      3    29
                                    Male        0    11
               (49375.0, 83750.0]   Female      2    24
                                    Male        2    11
               (83750.0, 118125.0]  Female      3    23
                                    Male        0    18
               (118125.0, 152500.0] Female      2    11
                                    Male        0    12

In [ ]:
Compras_etinge["suma"]=Compras_etinge[False]+Compras_etinge[True]

In [ ]:
Compras_etinge["% de compra"]=(Compras_etinge[False]/Compras_etinge["suma"])*100

In [ ]:
Compras_etinge["% de no compra"]=(Compras_etinge[True]/Compras_etinge["suma"])*100

In [ ]:
Compras_etinge

purchased                                   False  True  suma  % de compra  \
age_2          intervalos wage      gender                                   
(16.954, 28.5] (14862.5, 49375.0]   Female     28     0    28   100.000000   
                                    Male       20     0    20   100.000000   
               (49375.0, 83750.0]   Female     46     2    48    95.833333   
                                    Male       50     0    50   100.000000   
               (83750.0, 118125.0]  Female     12     1    13    92.307692   
                                    Male       15     0    15   100.000000   
               (118125.0, 152500.0] Female      1     1     2    50.000000   
                                    Male        0     3     3     0.000000   
(28.5, 40.0]   (14862.5, 49375.0]   Female     21     0    21   100.000000   
                                    Male       31     0    31   100.000000   
               (49375.0, 83750.0]   Female    104    13   117    88.888889   
                                    Male      112    17   129    86.821705   
               (83750.0, 118125.0]  Female      9     8    17    52.941176   
                                    Male        8    17    25    32.000000   
               (118125.0, 152500.0] Female      1    18    19     5.263158   
                                    Male        0    14    14     0.000000   
(40.0, 51.5]   (14862.5, 49375.0]   Female     11    27    38    28.947368   
                                    Male        4    29    33    12.121212   
               (49375.0, 83750.0]   Female     48    24    72    66.666667   
                                    Male       53    21    74    71.621622   
               (83750.0, 118125.0]  Female      6    19    25    24.000000   
                                    Male        0    20    20     0.000000   
               (118125.0, 152500.0] Female      4    21    25    16.000000   
                                    Male        2     8    10    20.000000   
(51.5, 63.0]   (14862.5, 49375.0]   Female      3    29    32     9.375000   
                                    Male        0    11    11     0.000000   
               (49375.0, 83750.0]   Female      2    24    26     7.692308   
                                    Male        2    11    13    15.384615   
               (83750.0, 118125.0]  Female      3    23    26    11.538462   
                                    Male        0    18    18     0.000000   
               (118125.0, 152500.0] Female      2    11    13    15.384615   
                                    Male        0    12    12     0.000000   

purchased                                   % de no compra  
age_2          intervalos wage      gender                  
(16.954, 28.5] (14862.5, 49375.0]   Female        0.000000  
                                    Male          0.000000  
               (49375.0, 83750.0]   Female        4.166667  
                                    Male          0.000000  
               (83750.0, 118125.0]  Female        7.692308  
                                    Male          0.000000  
               (118125.0, 152500.0] Female       50.000000  
                                    Male        100.000000  
(28.5, 40.0]   (14862.5, 49375.0]   Female        0.000000  
                                    Male          0.000000  
               (49375.0, 83750.0]   Female       11.111111  
                                    Male         13.178295  
               (83750.0, 118125.0]  Female       47.058824  
                                    Male         68.000000  
               (118125.0, 152500.0] Female       94.736842  
                                    Male        100.000000  
(40.0, 51.5]   (14862.5, 49375.0]   Female       71.052632  
                                    Male         87.878788  
               (49375.0, 83750.0]   Female       33.333333  
                                    Male         28.378378  
     

Con la realización de esta tabla pivot podemos observar: En primer lugar, exite una relacion inversa entre la edad y la probabilidad de haber comprado, viendose esto reflejado en que los hombres jovenes, y sin distincion de ingreso, suelen tener alto pocentaje de compra; mientras que los hombres mayores, y sin distincion de ingresos, suelen tener bajos porcentajes de compra.

Ingreso medio y mediano por grupos etarios

In [ ]:
data.pivot_table(index=["age_2"],
                  values=["wage"],
                  aggfunc=["mean","median"])

mean        median
                        wage          wage
age_2                                     
(16.954, 28.5]  60784.415330  60500.000000
(28.5, 40.0]    74028.958470  72000.000000
(40.0, 51.5]    73847.994500  73500.000000
(51.5, 63.0]    80467.489464  77168.831169

Podemos evidenciar que existe un sesgo positivo en cuanto al salario en todos los grupos etarios, existiendo datos muy altos que desplazan la media hacia la derecha.

Ingreso medio y mediano por género

In [ ]:
data.pivot_table(index=["gender"],
                  values=["wage"],
                  aggfunc=["mean","median"])

mean        median
                wage          wage
gender                            
Female  74072.397544  76953.846154
Male    70943.233496  71471.518987

Podemos evidenciar que existe un sesgo negativo en cuanto al salario, es decir que existen datos muy pequeños que desplazan la media hacia la izquierda, ademas se puede observar que la magnitud de este sesgo es mayor  en mujeres que en hombres.



**Conclusion General**



En treminos generales la limpieza de la base de datos fue bastante sencilla debido a que presentaba una unica variable con datos nulos, adicionalmente el tema de outliers era pertinente analizarlo igualmente en una sola variable, por lo que se tenia que utilizar un tratamiento univariado que permitio eliminar este tipo de datos de forma simple a traves de los metodos indicados. Asi mismo se realizo un enriquecimiento del dataset que permitio hacer una clasificacion de datos exhaustiva la cual permitio visualizar de forma clara los datos por medio de ntervalos y tablas pivot 

